File: analytics_server.py
from flask import Flask, request, jsonify, make_response
from flask_cors import CORS
import sqlite3
import datetime
import os
import json

app = Flask(__name__)
# Allow CORS for all domains for now, tighten in production
CORS(app, resources={r"/api/*": {"origins": "*"}})

DB_PATH = 'analytics.db'

def init_db():
    conn = sqlite3.connect(DB_PATH)
    c = conn.cursor()
    c.execute('''
        CREATE TABLE IF NOT EXISTS events (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            timestamp TEXT,
            visitor_id TEXT,
            session_id TEXT,
            event_type TEXT,
            data TEXT,
            ip_address TEXT,
            country TEXT,
            city TEXT,
            user_agent TEXT
        )
    ''')
    conn.commit()
    conn.close()

init_db()

@app.route('/api/analytics', methods=['POST', 'OPTIONS'])
def collect_analytics():
    if request.method == 'OPTIONS':
        return _build_cors_preflight_response()

    try:
        # Handle both JSON body and beacon API (sometimes sends distinct types)
        if request.is_json:
            payload = request.get_json()
        else:
            # Fallback for sendBeacon with Blob
            payload = json.loads(request.data)
            
        visitor_ip = request.remote_addr
        user_agent = request.headers.get('User-Agent')
        
        # Extract location data if provided by frontend (or use server-side geoip in future)
        location = payload.get('location', {})
        
        conn = sqlite3.connect(DB_PATH)
        c = conn.cursor()
        c.execute('''
            INSERT INTO events 
            (timestamp, visitor_id, session_id, event_type, data, ip_address, country, city, user_agent)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (
            payload.get('timestamp', datetime.datetime.now().isoformat()),
            payload.get('visitorId'),
            payload.get('sessionId'),
            payload.get('event'),
            json.dumps(payload.get('data', {})),
            visitor_ip,
            location.get('country', 'Unknown'),
            location.get('city', 'Unknown'),
            user_agent
        ))
        conn.commit()
        conn.close()
        
        return jsonify({"status": "success"}), 201
        
    except Exception as e:
        print(f"Error saving analytics: {e}")
        return jsonify({"status": "error", "message": str(e)}), 500

@app.route('/api/stats', methods=['GET'])
def get_stats():
    # Simple auth check (implement real auth in production!)
    auth_header = request.headers.get('Authorization')
    if auth_header != 'Bearer RECON_PRO_ADMIN_SECRET':
        return jsonify({"error": "Unauthorized"}), 401

    try:
        conn = sqlite3.connect(DB_PATH)
        conn.row_factory = sqlite3.Row
        c = conn.cursor()
        
        # Get total visits (unique sessions)
        c.execute("SELECT COUNT(DISTINCT session_id) as count FROM events WHERE event_type = 'page_view'")
        total_visits = c.fetchone()['count']
        
        # Get total unique visitors
        c.execute("SELECT COUNT(DISTINCT visitor_id) as count FROM events")
        total_visitors = c.fetchone()['count']
        
        # Get events per type
        c.execute("SELECT event_type, COUNT(*) as count FROM events GROUP BY event_type")
        event_distribution = {row['event_type']: row['count'] for row in c.fetchall()}
        
        # Get recent events
        c.execute("SELECT * FROM events ORDER BY id DESC LIMIT 50")
        recent_events = [dict(row) for row in c.fetchall()]
        
        conn.close()
        
        return jsonify({
            "summary": {
                "total_visits": total_visits,
                "total_visitors": total_visitors,
                "event_distribution": event_distribution
            },
            "recent_events": recent_events
        })
        
    except Exception as e:
        return jsonify({"error": str(e)}), 500

def _build_cors_preflight_response():
    response = make_response()
    response.headers.add("Access-Control-Allow-Origin", "*")
    response.headers.add("Access-Control-Allow-Headers", "*")
    response.headers.add("Access-Control-Allow-Methods", "*")
    return response

if __name__ == '__main__':
    # Run slightly differently for dev vs prod (gunicorn)
    app.run(host='0.0.0.0', port=5000)

File: analytics_server.php
<?php
/**
 * RECON PRO - PHP Analytics Backend
 * Compatible with Shared Hosting (Hostinger, cPanel, etc.)
 * Drop this file into the same directory as your index.html
 */

header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, GET, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');
header('Content-Type: application/json');

// Database Configuration (SQLite)
$dbFile = __DIR__ . '/analytics.db';

// Initialize Database if not exists
if (!file_exists($dbFile)) {
    try {
        $db = new PDO('sqlite:' . $dbFile);
        $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $db->exec("CREATE TABLE IF NOT EXISTS events (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            timestamp TEXT,
            visitor_id TEXT,
            session_id TEXT,
            event_type TEXT,
            data TEXT,
            ip_address TEXT,
            country TEXT,
            city TEXT,
            user_agent TEXT
        )");
        chmod($dbFile, 0664); // Ensure writable
    } catch (PDOException $e) {
        http_response_code(500);
        die(json_encode(['status' => 'error', 'message' => 'Database initialization failed: ' . $e->getMessage()]));
    }
}

// Handle Requests
$method = $_SERVER['REQUEST_METHOD'];

if ($method === 'OPTIONS') {
    http_response_code(200);
    exit();
}

if ($method === 'POST') {
    // Collect Data
    $input = file_get_contents('php://input');
    $payload = json_decode($input, true);

    if (!$payload) {
        http_response_code(400);
        die(json_encode(['status' => 'error', 'message' => 'Invalid JSON']));
    }

    $ip = $_SERVER['REMOTE_ADDR'];
    $ua = $_SERVER['HTTP_USER_AGENT'] ?? 'Unknown';
    $location = $payload['location'] ?? [];

    try {
        $db = new PDO('sqlite:' . $dbFile);
        $stmt = $db->prepare("INSERT INTO events (timestamp, visitor_id, session_id, event_type, data, ip_address, country, city, user_agent) VALUES (:ts, :vid, :sid, :evt, :dat, :ip, :cnt, :cty, :ua)");

        $stmt->execute([
            ':ts' => $payload['timestamp'] ?? date('c'),
            ':vid' => $payload['visitorId'] ?? null,
            ':sid' => $payload['sessionId'] ?? null,
            ':evt' => $payload['event'] ?? 'unknown',
            ':dat' => json_encode($payload['data'] ?? []),
            ':ip' => $ip,
            ':cnt' => $location['country'] ?? 'Unknown',
            ':cty' => $location['city'] ?? 'Unknown',
            ':ua' => $ua
        ]);

        http_response_code(201);
        echo json_encode(['status' => 'success']);
    } catch (PDOException $e) {
        http_response_code(500);
        echo json_encode(['status' => 'error', 'message' => 'Write failed']);
    }
    exit();
}

if ($method === 'GET') {
    // Admin Stats Endpoint
    // SECURITY: Change this token!
    $secret = 'afsdlj24356wneiv346wpeomafpll2092342lkajsdf';
    $headers = getallheaders();
    $auth = $headers['Authorization'] ?? '';

    if ($auth !== "Bearer $secret") {
        http_response_code(401);
        die(json_encode(['error' => 'Unauthorized']));
    }

    try {
        $db = new PDO('sqlite:' . $dbFile);

        $stmt = $db->query("SELECT COUNT(DISTINCT session_id) as count FROM events WHERE event_type = 'page_view'");
        $totalVisits = $stmt->fetch(PDO::FETCH_ASSOC)['count'];

        // Total Visitors
        $stmt = $db->query("SELECT COUNT(DISTINCT visitor_id) as count FROM events");
        $totalVisitors = $stmt->fetch(PDO::FETCH_ASSOC)['count'];

        // Event Distribution
        $stmt = $db->query("SELECT event_type, COUNT(*) as count FROM events GROUP BY event_type");
        $dist = $stmt->fetchAll(PDO::FETCH_KEY_PAIR);

        // Recent Events
        $stmt = $db->query("SELECT * FROM events ORDER BY id DESC LIMIT 50");
        $recent = $stmt->fetchAll(PDO::FETCH_ASSOC);

        echo json_encode([
            'summary' => [
                'total_visits' => $totalVisits,
                'total_visitors' => $totalVisitors,
                'event_distribution' => $dist
            ],
            'recent_events' => $recent
        ]);

    } catch (PDOException $e) {
        http_response_code(500);
        echo json_encode(['error' => $e->getMessage()]);
    }
    exit();
}
?>